taraxa/state/state_transition/state_transition.go

```diff
@@ -93,18 +104,29 @@ func (self *StateTransition) evm_state_checkpoint() {
 func (self *StateTransition) BeginBlock(blk_info *vm.BlockInfo) {
 	self.begin_block()
 	blk_n := self.pending_blk_state.GetNumber()
-	rules_changed := self.evm.SetBlock(&vm.Block{blk_n, *blk_info}, self.chain_cfg.ETHChainConfig.Rules(blk_n))
+	rules_changed := self.evm.SetBlock(&vm.Block{blk_n, *blk_info}, self.chain_config.ETHChainConfig.Rules(blk_n))
 	if self.dpos_contract != nil && rules_changed {
 		self.dpos_contract.Register(self.evm.RegisterPrecompiledContract)
 	}
-	if self.chain_cfg.ETHChainConfig.IsDAOFork(blk_n) {
+	if self.chain_config.ETHChainConfig.IsDAOFork(blk_n) {
 		misc.ApplyDAOHardFork(&self.evm_state)
 		self.evm_state_checkpoint()
 	}
+	if self.chain_config.Hardforks.IsFixGenesisFork(blk_n) {
+		if self.new_chain_config == nil {
+			panic("we should have new_chain_config for hardfork")
+		}
+		self.chain_config = self.new_chain_config
+		self.new_chain_config = nil
+
+		self.dpos_contract.ResetGenesisAddresses(self.chain_config.DPOS.GenesisState)
+		self.dpos_contract.UpdateConfig(*self.chain_config.DPOS)
+		self.chain_config.Hardforks.ApplyFixGenesisHardfork(self.chain_config.GenesisBalances, self.chain_config.DPOS, &self.evm_state, self.dpos_contract, self.get_reader(blk_n-1))
```

Comment:
calling actual hf function on self.chain_config.Hardforks does not seem ok. Config is config, it shouldnt really apply any hardfork logic